File: angular-spring-boot/backend/src/main/java/ru/isnagornov/templates/rest/AbstractController.java
package ru.isnagornov.templates.rest;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import ru.isnagornov.templates.mapper.CommonMapper;
import ru.isnagornov.templates.service.AbstractService;

import java.util.List;

public abstract class AbstractController<T, M extends CommonMapper<T>, S extends AbstractService<T, M>> {

    protected S service;

    public AbstractController(S service) {
        this.service = service;
    }

    @PostMapping("/add")
    public ResponseEntity<HttpStatus> add(@RequestBody T entity) {
        this.service.add(entity);

        return new ResponseEntity<>(HttpStatus.OK);
    }

    @GetMapping(path = "/{id}")
    public T getById(@PathVariable("id") Long id) {
        return this.service.getById(id);
    }

    @GetMapping("/getAll")
    public List<T> getAll() {
        return this.service.getAll();
    }


    @PutMapping(path = "/update")
    public ResponseEntity<HttpStatus> update(@RequestBody T entity) {
        this.service.update(entity);

        return new ResponseEntity<>(HttpStatus.OK);
    }

    @DeleteMapping(path = "/{id}")
    public ResponseEntity<HttpStatus> delete(@PathVariable("id") Long id) {
        this.service.delete(id);

        return new ResponseEntity<>(HttpStatus.OK);
    }

}

File: angular-spring-boot-kotlin/backend/src/main/kotlin/ru/isnagornov/templates/rest/AbstractController.kt
package ru.isnagornov.templates.rest

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import ru.isnagornov.templates.mapper.CommonMapper
import ru.isnagornov.templates.service.AbstractService

abstract class AbstractController<T, M : CommonMapper<T>, S : AbstractService<T, M>>(protected var service: S) {

    val all: List<T>
        @GetMapping("/getAll")
        get() = this.service.all

    @PostMapping("/add")
    fun add(@RequestBody entity: T): ResponseEntity<HttpStatus> {
        this.service.add(entity)

        return ResponseEntity(HttpStatus.OK)
    }

    @GetMapping(path = ["/{id}"])
    fun getById(@PathVariable("id") id: Long?): T {
        return this.service.getById(id)
    }


    @PutMapping(path = ["/update"])
    fun update(@RequestBody entity: T): ResponseEntity<HttpStatus> {
        this.service.update(entity)

        return ResponseEntity(HttpStatus.OK)
    }

    @DeleteMapping(path = ["/{id}"])
    fun delete(@PathVariable("id") id: Long?): ResponseEntity<HttpStatus> {
        this.service.delete(id)

        return ResponseEntity(HttpStatus.OK)
    }

}
